File: clients/go/mongocore/cursor.go
package mongocore

import (
	"context"
	"io"

	pb "github.com/rozza/mongocore/clients/go/proto"
	"go.mongodb.org/mongo-driver/v2/bson"
)

// batchStream is the interface shared by both FindStream and AggregateStream clients.
// Both return DocumentBatch messages via Recv().
type batchStream interface {
	Recv() (*pb.DocumentBatch, error)
}

// Cursor iterates over documents from a streaming gRPC call.
// The underlying RPC is called on first Next() invocation (lazy).
// Must be closed when done to release resources.
type Cursor struct {
	stream   batchStream
	cancelFn context.CancelFunc
	buffer   []bson.D
	index    int
	done     bool
	err      error

	// Lazy init fields
	initFn func(ctx context.Context) (batchStream, context.CancelFunc, error)
}

// Next advances the cursor to the next document.
// Returns true if a document is available via Doc(), false when exhausted or on error.
func (c *Cursor) Next(ctx context.Context) bool {
	if c.err != nil || c.done {
		return false
	}

	// Lazy initialization
	if c.stream == nil {
		stream, cancel, err := c.initFn(ctx)
		if err != nil {
			c.err = err
			return false
		}
		c.stream = stream
		c.cancelFn = cancel
	}

	// Try buffer first
	if c.index < len(c.buffer) {
		return true
	}

	// Fetch next batch via the batchStream interface
	batch, err := c.stream.Recv()
	if err != nil {
		if err == io.EOF {
			c.done = true
		} else {
			c.err = err
		}
		return false
	}

	c.buffer = make([]bson.D, 0, len(batch.Documents))
	for _, d := range batch.Documents {
		doc, err := decodeBsonDoc(d.Data)
		if err != nil {
			c.err = err
			return false
		}
		c.buffer = append(c.buffer, doc)
	}
	c.index = 0

	if !batch.HasMore {
		c.done = true
	}

	return len(c.buffer) > 0
}

// Doc returns the current document. Must only be called after Next() returns true.
func (c *Cursor) Doc() bson.D {
	doc := c.buffer[c.index]
	c.index++
	return doc
}

// Err returns any error that occurred during iteration.
func (c *Cursor) Err() error {
	return c.err
}

// Close cancels the underlying stream and releases resources.
func (c *Cursor) Close() error {
	if c.cancelFn != nil {
		c.cancelFn()
	}
	c.done = true
	return nil
}

// All collects all remaining documents into a slice.
func (c *Cursor) All(ctx context.Context) ([]bson.D, error) {
	var results []bson.D
	for c.Next(ctx) {
		results = append(results, c.Doc())
	}
	if c.err != nil {
		return nil, c.err
	}
	return results, nil
}

// Ensure Cursor implements io.Closer.
var _ io.Closer = (*Cursor)(nil)

File: clients/python/src/mongocore/cursor.py
"""Async cursor over streaming gRPC query results."""


_CLIENT_METADATA = [("x-client-language", "python")]


class Cursor:
    """Async iterator that yields documents from a streaming gRPC call.

    The underlying RPC is not called until iteration begins (lazy).
    """

    def __init__(self, stub, request, rpc_method: str, decode_fn):
        self._stub = stub
        self._request = request
        self._rpc_method = rpc_method
        self._decode_fn = decode_fn
        self._stream = None
        self._buffer: list = []
        self._buffer_index: int = 0
        self._exhausted: bool = False

    def __aiter__(self):
        return self

    async def __anext__(self):
        if self._buffer_index < len(self._buffer):
            doc = self._buffer[self._buffer_index]
            self._buffer_index += 1
            return doc

        if self._exhausted:
            raise StopAsyncIteration

        await self._fetch_next_batch()

        if self._buffer_index < len(self._buffer):
            doc = self._buffer[self._buffer_index]
            self._buffer_index += 1
            return doc

        raise StopAsyncIteration

    async def _fetch_next_batch(self):
        if self._stream is None:
            rpc = getattr(self._stub, self._rpc_method)
            self._stream = rpc(self._request, metadata=_CLIENT_METADATA)

        try:
            batch = await self._stream.read()
            if batch is None:
                self._exhausted = True
                return
        except Exception:
            self._exhausted = True
            raise

        self._buffer = [self._decode_fn(doc.data) for doc in batch.documents]
        self._buffer_index = 0

        if not batch.has_more:
            self._exhausted = True

    async def to_list(self) -> list:
        """Collect all documents into a list."""
        results = []
        async for doc in self:
            results.append(doc)
        return results

    async def close(self):
        """Cancel the underlying stream."""
        if self._stream is not None:
            self._stream.cancel()
            self._stream = None
        self._exhausted = True
